frame/staking/src/lib.rs

```diff
@@ -819,8 +896,49 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
-			Self::ensure_storage_upgraded();
+		fn on_initialize(now: T::BlockNumber) {
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::election_status() == OffchainElectionStatus::None &&
+				// and an era is about to be changed.
+				Self::is_current_session_final()
+			{
+				// TODO: maybe we can be naive like im-online and just assume block == slot?
+				let next_session_change =
+					T::NextSessionChange::predict_next_session_change(now);
+				if let Some(remaining) = next_session_change.checked_sub(&now) {
+					if remaining <= T::ElectionLookahead::get() && !remaining.is_zero() {
+						// Set the flag to make sure we don't waste any compute here in the same era
+						// after we have triggered the offline compute.
+						<ElectionStatus<T>>::put(
+							OffchainElectionStatus::<T::BlockNumber>::Triggered(now)
+						);
+						frame_support::print("detected a good block to trigger offchain worker.");
```

Comment:
I think that should be just a debug log messages instead of `print`.

Reply:
most prints in the code are for testing now, since some of the logic is missing. Will cleanup. 👍